src/material/core/datetime/native-date-adapter.ts

```diff
@@ -257,4 +322,47 @@ export class NativeDateAdapter extends DateAdapter<Date> {
     d.setUTCHours(date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds());
     return dtf.format(d);
   }
+
+  private _nativeParseFallback(value: string): Date {
+    const date = new Date(Date.parse(value));
+    if (!this.isValid(date)) {
+      return date;
+    }
+
+    // Native parsing sometimes assumes UTC, sometimes does not.
+    // We have to remove the difference between the two in order to get the date as a local date.
+
+    const compareDate = new Date(date.getFullYear(), date.getMonth(), date.getDate(), 0, 0, 0);
+    const difference = date.getTime() - compareDate.getTime();
+    if (difference === 0) {
+      return date;
+    }
+
+    return new Date(
+      date.getUTCFullYear(),
+      date.getUTCMonth(),
+      date.getUTCDate(),
+      date.getUTCHours(),
+      date.getUTCMinutes(),
+      date.getUTCSeconds(),
+      date.getUTCMilliseconds(),
+    );
+  }
+
+  private _dateComponentsAreValid(year: number, month: number, day: number) {
+    if (year < 0 || year > 9999 || month < 0 || month > 11 || day < 1 || day > 31) {
+      return false;
+    }
+
+    if (month === 1) {
```

Comment:
an alternative instead of writing all of this logic could be to just do:

```
const result = createDate(year, month, day);
if (result.getFullYear() === year && result.getMonth() === month && result.getDate() === day) {
  ...
} else {
  ...
}
```